File: gauss_implementation/code.rs
use rand::Rng;
use std::env;
use std::time::Instant;

const MAXN: usize = 2000;

type Matrix = Vec<Vec<f64>>;
type Vector = Vec<f64>;

fn gauss(a: &mut Matrix, b: &mut Vector, n: usize) -> Vector {
    let mut x = vec![0.0; n];

    for norm in 0..n - 1 {
        for row in norm + 1..n {
            let multiplier = a[row][norm] / a[norm][norm];
            for col in norm..n {
                a[row][col] -= a[norm][col] * multiplier;
            }
            b[row] -= b[norm] * multiplier;
        }
    }

    for row in (0..n).rev() {
        x[row] = b[row];
        for col in row + 1..n {
            x[row] -= a[row][col] * x[col];
        }
        x[row] /= a[row][row];
    }

    x
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <matrix_size>", args[0]);
        return;
    }

    let n: usize = args[1].parse().unwrap_or(0);
    if n < 1 || n > MAXN {
        eprintln!("Matrix size must be between 1 and {}", MAXN);
        return;
    }

    let mut rng = rand::thread_rng();
    let mut a: Matrix = (0..n)
        .map(|_| (0..n).map(|_| rng.gen_range(0.0..100.0)).collect())
        .collect();
    let mut b: Vector = (0..n).map(|_| rng.gen_range(0.0..100.0)).collect();

    let start = Instant::now();
    gauss(&mut a, &mut b, n);
    let elapsed = start.elapsed();

    println!("Elapsed time: {:.6} seconds", elapsed.as_secs_f64());
}

File: gauss_implementation/code.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAXN 2000

void gauss(double A[MAXN][MAXN], double B[MAXN], double X[MAXN], int N) {
    for (int norm = 0; norm < N - 1; norm++) {
        for (int row = norm + 1; row < N; row++) {
            double multiplier = A[row][norm] / A[norm][norm];
            for (int col = norm; col < N; col++) {
                A[row][col] -= A[norm][col] * multiplier;
            }
            B[row] -= B[norm] * multiplier;
        }
    }
    for (int row = N - 1; row >= 0; row--) {
        X[row] = B[row];
        for (int col = N - 1; col > row; col--) {
            X[row] -= A[row][col] * X[col];
        }
        X[row] /= A[row][row];
    }
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <matrix_size>\n", argv[0]);
        return -1;
    }
    int N = atoi(argv[1]);
    if (N < 1 || N > MAXN) {
        printf("Matrix size must be between 1 and %d\n", MAXN);
        return -1;
    }
    
    double A[MAXN][MAXN], B[MAXN], X[MAXN];
    srand(time(NULL));
    for (int i = 0; i < N; i++) {
        B[i] = rand() % 100;
        for (int j = 0; j < N; j++) {
            A[i][j] = rand() % 100;
        }
    }
    
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    
    gauss(A, B, X, N);
    
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    printf("Elapsed time: %.6f seconds\n", elapsed);
    
    return 0;
}
